src/components/Tables/Filters.tsx

```diff
@@ -0,0 +1,189 @@
+import React from "react";
+import styled from "styled-components";
+import { TableHeaders } from "@/components/Tables/Table";
+import { DatePicker } from "@mui/x-date-pickers";
+import FreeFormTextInput from "../DataInput/FreeFormTextInput";
+
+export interface Filter<Data, State> {
+    shouldFilter: (data: Data, state: State) => boolean;
+    filterComponent: (state: State, setState: (state: State) => void) => React.ReactElement;
+    state: State;
+    initialState: State;
+    clear: (state: State, setState: (state: State) => void) => void;
+}
+
+export interface KeyedFilter<Data, Key extends keyof Data, State> extends Filter<Data, State> {
+    key: Key;
+    label: string;
+}
+
+export const isKeyedFilter = <Data, Key extends keyof Data, State>(
+    filter: Filter<Data, State>
+): filter is KeyedFilter<Data, Key, State> => {
+    return "key" in filter;
+};
+
+const keyedFilter = <Data, Key extends keyof Data, State>(
+    key: Key,
+    label: string,
+    filter: Filter<Data, State>
+): KeyedFilter<Data, Key, State> => {
+    return {
+        key,
+        label,
+        ...filter,
+    };
+};
+
+export const headerLabelFromKey = <Data, Key extends keyof Data>(
+    headers: TableHeaders<Data>,
+    key: Key
+): string => {
+    return headers.find(([headerKey]) => headerKey === key)?.[1] ?? key.toString();
+};
+
+const TextFilterStyling = styled.div`
+    & input[type="text"].MuiInputBase-input.MuiOutlinedInput-input {
+        border: none;
+    }
+`;
+
+interface TextFilterProps<Data, Key extends keyof Data> {
+    key: Key;
+    label: string;
+    caseSensitive?: boolean;
+    initialValue?: string;
+}
+
+const defaultToString = (value: unknown): string => {
+    if (typeof value === "string") {
+        return value;
+    }
+
+    return JSON.stringify(value);
+};
+
+export const textFilter = <Data, Key extends keyof Data>({
+    key,
+    label,
+    caseSensitive = false,
+    initialValue = "",
+}: TextFilterProps<Data, Key>): KeyedFilter<Data, Key, string> => {
+    return keyedFilter(key, label, {
+        state: initialValue,
+        initialState: initialValue,
+        shouldFilter: (data, state) => {
+            let string = defaultToString(data[key]);
+
+            if (!caseSensitive) {
+                string = string.toLowerCase();
+                state = state.toLowerCase();
```

Comment:
I'm surprised it lets you set `state` without a `let` explicitly, and just 'setting' the input variable